odl/discr/discr_utils.py

```diff
@@ -0,0 +1,812 @@
+        A single string is interpreted as a global scheme for all
+        axes.
+    nn_variants : string or sequence of strings, optional
+        Which variant ('left' or 'right') to use in nearest neighbor
+        interpolation for which axis. A single string is interpreted
+        as a global variant for all axes.
+        This option has no effect for schemes other than nearest
+        neighbor.
+    Examples
+    --------
+    Choose linear interpolation in the first axis and nearest neighbor in
+    the second:
+    >>> part = odl.uniform_partition([0, 0], [1, 5], shape=(2, 4))
+    >>> part.coord_vectors
+    (array([ 0.25,  0.75]), array([ 0.625,  1.875,  3.125,  4.375]))
+    >>> f = np.array([[1, 2, 3, 4],
+    ...               [5, 6, 7, 8]],
+    ...              dtype=float)
+    >>> interpolator = per_axis_interpolator(
+    ...     f, part.coord_vectors, ['linear', 'nearest']
+    ... )
+    >>> interpolator([1, 1])  # single point
+    2.5
+    >>> interpolator([[0.5, 2.0],
+    ...               [0.0, 4.5],
+    ...               [0.0, 3.0]])  # 3 points at once
+    array([ 6. , -7.5,  1.5])
+    >>> from odl.discr.grid import sparse_meshgrid
+    >>> mesh = sparse_meshgrid([0.0, 0.5, 1.0], [1.5, 3.5])
+    >>> interpolator(mesh)  # 3x2 grid of points
+    array([[ 1. ,  1.5],
+           [ 4. ,  5. ],
+           [ 3. ,  3.5]])
+    """
+    f = np.asarray(f)
+
+    schemes_in = schemes
+    if is_string(schemes):
+        scheme = str(schemes).lower()
+        if scheme not in _SUPPORTED_INTERP_SCHEMES:
+            raise ValueError('`schemes` {!r} not understood'
+                             ''.format(schemes_in))
+        schemes = [scheme] * f.ndim
+    else:
+        schemes = [str(scm).lower() if scm is not None else None
+                   for scm in schemes]
+
+    nn_variants_in = nn_variants
+    if nn_variants is None:
+        nn_variants = ['left' if scm == 'nearest' else None
+                       for scm in schemes]
+    else:
+        if is_string(nn_variants):
+            # Make list with `nn_variants` where `schemes == 'nearest'`,
+            # else `None` (variants only applies to axes with nn
+            # interpolation)
+            nn_variants = [nn_variants if scm == 'nearest' else None
+                           for scm in schemes]
+            if str(nn_variants_in).lower() not in ('left', 'right'):
+                raise ValueError('`nn_variants` {!r} not understood'
+                                 ''.format(nn_variants_in))
+        else:
+            nn_variants = [str(var).lower() if var is not None else None
+                           for var in nn_variants]
+
+    for i in range(f.ndim):
+        # Reaching a raise condition here only happens for invalid
+        # sequences of inputs, single-input case has been checked above
+        if schemes[i] not in _SUPPORTED_INTERP_SCHEMES:
+            raise ValueError('`interp[{}]={!r}` not understood'
```

Comment:
Give allowed values in error message? Also below

Reply:
Done